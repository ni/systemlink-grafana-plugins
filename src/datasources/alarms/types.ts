import { DataQuery } from '@grafana/schema';

export interface AlarmsQuery extends DataQuery {}
