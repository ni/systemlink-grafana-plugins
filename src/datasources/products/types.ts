import { DataQuery } from '@grafana/schema'

export interface ProductQuery extends DataQuery {
  properties?: Properties[];
  orderBy?: string;
  descending?: boolean;
  recordCount?: number;
}

export enum Properties {
  id = 'ID',
  partNumber = 'PART_NUMBER',
  name = 'NAME',
  family = 'FAMILY',
  updatedAt = 'UPDATED_AT',
  workspace = 'WORKSPACE',
  keywords = 'KEYWORDS',
  properties = 'PROPERTIES',
  fileIds = 'FILE_IDS',
}

export const PropertiesOptions = {
  ID: 'id',
  PART_NUMBER: 'partNumber',
  NAME: 'name',
  FAMILY: 'family',
  UPDATEDAT: 'updatedAt',
  WORKSPACE: 'workspace',
  KEYWORDS: 'keywords',
  PROPERTIES: 'properties',
  FILE_IDS: 'fileIds',
};

export const OrderBy = [
  {
    value: 'ID',
    label: 'ID',
    description: `ID of the product`,
  },
  {
    value: 'PART_NUMBER',
    label: 'Part Number',
    description: `Part number of the product`,
  },
  {
    value: 'NAME',
    label: 'Product Name',
    description: `Name of the product`,
  },
  {
    value: 'FAMILY',
    label: 'Family',
    description: `Family of the product`,
  },
  {
    value: 'WORKSPACE',
    label: 'Workspace',
    description: `Workspace of the product`,
  },
  {
    value: 'UPDATED_AT',
    label: 'Updated At',
    description: `Latest update at time of the product`,
  }
];

export interface QueryProductResponse {
  products: ProductResponseProperties[],
  continuationToken: string,
  totalCount: number
}

export interface ProductResponseProperties {
  id: string;
  partNumber: string;
  name?: string;
  family?: string;
  updatedAt?: string;
  workspace?: string;
  keywords?: any;
  properties?: Object;
  fileIds?: string[];
  returnCount?: number;
}
