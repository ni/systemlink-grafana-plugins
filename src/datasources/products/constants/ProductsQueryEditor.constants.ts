export const TAKE_LIMIT = 10000;

export const recordCountErrorMessages = {
    greaterOrEqualToZero: 'Enter a value greater than or equal to 0',
    lessOrEqualToTakeLimit: `Enter a value less than or equal to ${TAKE_LIMIT.toLocaleString()}`,
};
