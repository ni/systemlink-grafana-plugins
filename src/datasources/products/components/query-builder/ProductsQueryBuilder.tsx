import { useTheme2 } from "@grafana/ui";
import { queryBuilderMessages, QueryBuilderOperations } from "core/query-builder.constants";
import { expressionBuilderCallback, expressionReaderCallback } from "core/query-builder.utils";
import { Workspace, QueryBuilderOption } from "core/types";
import { filterXSSField, filterXSSLINQExpression } from "core/utils";
import { ProductsQueryBuilderFields } from "datasources/products/constants/ProductsQueryBuilder.constants";
import { QBField } from "datasources/products/types";
import React, { useState, useEffect, useMemo } from "react";
import QueryBuilder, { QueryBuilderCustomOperation, QueryBuilderProps } from "smart-webcomponents-react/querybuilder";

type ProductsQueryBuilderProps = QueryBuilderProps & React.HTMLAttributes<Element> & {
  filter?: string;
  workspaces: Workspace[];
  partNumbers: string[];
  globalVariableOptions: QueryBuilderOption[];
  staticFields?: QBField[];
};

export const ProductsQueryBuilder: React.FC<ProductsQueryBuilderProps> = ({
  filter,
  onChange,
  workspaces,
  partNumbers,
  globalVariableOptions,
  staticFields
}) => {
  const theme = useTheme2();
  document.body.setAttribute("theme", theme.isDark ? "dark-orange" : "orange");

  const [fields, setFields] = useState<QBField[]>([]);
  const [operations, setOperations] = useState<QueryBuilderCustomOperation[]>([]);

  const sanitizedFilter = useMemo(() => {
    return filterXSSLINQExpression(filter);
  }, [filter])

  const workspaceField = useMemo(() => {
    const workspaceField = ProductsQueryBuilderFields.WORKSPACE;

    return {
      ...workspaceField,
      lookup: {
        ...workspaceField.lookup,
        dataSource: [
          ...(workspaceField.lookup?.dataSource || []),
          ...workspaces.map(({ id, name }) => ({ label: name, value: id })),
        ],
      },
    };
  }, [workspaces])

  const updatedAtField = useMemo(() => {
    const updatedField = ProductsQueryBuilderFields.UPDATEDAT;
    return {
      ...updatedField,
      lookup: {
        ...updatedField.lookup,
        dataSource: [
          ...(updatedField.lookup?.dataSource || []),
          { label: "From", value: "${__from:date}" },
          { label: "To", value: "${__to:date}" },
          { label: "Now", value: "${__now:date}" },
        ]
      }
    }
  }, []);

  const partNumberField = useMemo(() => {
    const partNumberField = ProductsQueryBuilderFields.PARTNUMBER;

    return {
      ...partNumberField,
      lookup: {
        ...partNumberField.lookup,
        dataSource: [
          ...(partNumberField.lookup?.dataSource || []),
          ...partNumbers.map((partNumber) => ({ label: partNumber, value: partNumber }))
        ]
      }
    }
  }, [partNumbers]);


  useEffect(() => {
    const updatedFields = [partNumberField, ...staticFields!, updatedAtField, workspaceField]
      .map((field) => {
        if (field.lookup?.dataSource) {
          return {
            ...field,
            lookup: {
              dataSource: [...globalVariableOptions, ...field.lookup!.dataSource].map(filterXSSField),
            },
          }
        }
        return field;
      });

    setFields(updatedFields);

    const options = Object.values(updatedFields).reduce((accumulator, fieldConfig) => {
      if (fieldConfig.lookup) {
        accumulator[fieldConfig.dataField!] = fieldConfig.lookup.dataSource;
      }

      return accumulator;
    }, {} as Record<string, QueryBuilderOption[]>);

    const callbacks = {
      expressionBuilderCallback: expressionBuilderCallback(options),
      expressionReaderCallback: expressionReaderCallback(options),
    };

    const customOperations = [
      QueryBuilderOperations.EQUALS,
      QueryBuilderOperations.DOES_NOT_EQUAL,
      QueryBuilderOperations.STARTS_WITH,
      QueryBuilderOperations.ENDS_WITH,
      QueryBuilderOperations.CONTAINS,
      QueryBuilderOperations.DOES_NOT_CONTAIN,
      QueryBuilderOperations.LESS_THAN,
      QueryBuilderOperations.LESS_THAN_OR_EQUAL_TO,
      QueryBuilderOperations.GREATER_THAN,
      QueryBuilderOperations.GREATER_THAN_OR_EQUAL_TO,
      QueryBuilderOperations.IS_BLANK,
      QueryBuilderOperations.IS_NOT_BLANK,
    ].map((operation) => {
      return {
        ...operation,
        ...callbacks,
      };
    });

    const keyValueOperations = [
      QueryBuilderOperations.KEY_VALUE_MATCH,
      QueryBuilderOperations.KEY_VALUE_DOES_NOT_MATCH,
      QueryBuilderOperations.KEY_VALUE_CONTAINS,
      QueryBuilderOperations.KEY_VALUE_DOES_NOT_CONTAINS,
      QueryBuilderOperations.KEY_VALUE_IS_GREATER_THAN,
      QueryBuilderOperations.KEY_VALUE_IS_GREATER_THAN_OR_EQUAL,
      QueryBuilderOperations.KEY_VALUE_IS_LESS_THAN,
      QueryBuilderOperations.KEY_VALUE_IS_LESS_THAN_OR_EQUAL,
      QueryBuilderOperations.KEY_VALUE_IS_NUMERICAL_EQUAL,
      QueryBuilderOperations.KEY_VALUE_IS_NUMERICAL_NOT_EQUAL
    ]

    setOperations([...customOperations, ...keyValueOperations]);

  }, [workspaceField, updatedAtField, partNumberField, staticFields, globalVariableOptions]);

  return (
    <QueryBuilder
      customOperations={operations}
      fields={fields}
      messages={queryBuilderMessages}
      onChange={onChange}
      value={sanitizedFilter}
      showIcons
    />
  );
}
