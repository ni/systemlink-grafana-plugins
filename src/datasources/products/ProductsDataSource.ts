import { DataFrameDTO, DataQueryRequest, DataSourceInstanceSettings, FieldType, LegacyMetricFindQueryOptions, MetricFindValue, TestDataSourceResponse } from '@grafana/data';
import { BackendSrv, TemplateSrv, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import { ProductQuery, ProductResponseProperties, ProductVariableQuery, Properties, PropertiesOptions, QueryProductResponse } from './types';
import { QueryBuilderOption, Workspace } from 'core/types';
import { parseErrorMessage } from 'core/errors';
import { ExpressionTransformFunction, transformComputedFieldsQuery } from 'core/query-builder.utils';
import { QueryBuilderOperations } from 'core/query-builder.constants';
import { ProductsQueryBuilderFieldNames } from './constants/ProductsQueryBuilder.constants';

export class ProductsDataSource extends DataSourceBase<ProductQuery> {
  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv()
  ) {
    super(instanceSettings, backendSrv, templateSrv);
    this.workspaceLoadedPromise = this.loadWorkspaces();
    this.partNumberLoadedPromise = this.getProductPartNumbers();
  }

  areWorkspacesLoaded$ = new Promise<void>(resolve => this.workspacesLoaded = resolve);
  arePartNumberLoaded$ = new Promise<void>(resolve => this.partNumberLoaded = resolve);
  error = '';

  baseUrl = this.instanceSettings.url + '/nitestmonitor';
  queryProductsUrl = this.baseUrl + '/v2/query-products';
  queryProductValuesUrl = this.baseUrl + '/v2/query-product-values';

  defaultQuery = {
    properties: [
      PropertiesOptions.PART_NUMBER,
      PropertiesOptions.NAME,
      PropertiesOptions.FAMILY,
      PropertiesOptions.WORKSPACE
    ] as Properties[],
    orderBy: Properties.updatedAt,
    descending: true,
    recordCount: 1000,
    queryBy: ''
  };

  readonly workspacesCache = new Map<string, Workspace>([]);
  readonly partNumbersCache = new Map<string, string>([]);
  readonly familyNamesCache = new Map<string, string>([]);

  readonly globalVariableOptions = (): QueryBuilderOption[] => this.getVariableOptions();

  private workspaceLoadedPromise: Promise<void>;
  private partNumberLoadedPromise: Promise<void>;

  private workspacesLoaded!: () => void;
  private partNumberLoaded!: () => void;

  async queryProducts(
    orderBy?: string,
    projection?: Properties[],
    filter?: string,
    take?: number,
    descending = false,
    returnCount = false
  ): Promise<QueryProductResponse> {
    try {
      const response = await this.post<QueryProductResponse>(this.queryProductsUrl, {
        filter,
        orderBy,
        descending,
        projection,
        take,
        returnCount
      });
      return response;
    } catch (error) {
      this.error = parseErrorMessage(error as Error)!;
      throw new Error(`An error occurred while querying products: ${this.error}`);
    }
  }

  async queryProductValues(fieldName: string): Promise<string[]> {
    try {
      return await this.post<string[]>(this.queryProductValuesUrl, {
        field: fieldName
      });
    } catch (error) {
      this.error = parseErrorMessage(error as Error)!;
      throw new Error(`An error occurred while querying product values: ${this.error}`);
    }
  }

  async runQuery(query: ProductQuery, options: DataQueryRequest): Promise<DataFrameDTO> {
    await this.workspaceLoadedPromise;
    await this.partNumberLoadedPromise;

    if (query.queryBy) {
      query.queryBy = transformComputedFieldsQuery(
        this.templateSrv.replace(query.queryBy, options.scopedVars),
        this.productsComputedDataFields,
      );
    }

    const products = (
      await this.queryProducts(
        query.orderBy,
        query.properties,
        query.queryBy,
        query.recordCount,
        query.descending
      )).products;

    if (products.length > 0) {
      const selectedFields = query.properties?.filter(
        (field: Properties) => Object.keys(products[0]).includes(field)) || [];
      const fields = selectedFields.map((field) => {
        const isTimeField = field === PropertiesOptions.UPDATEDAT;
        const fieldType = isTimeField
          ? FieldType.time
          : FieldType.string;

        const values = products
          .map(data => data[field as unknown as keyof ProductResponseProperties]);

        return {
          name: field,
          values: values.map(value => value != null
            ? (field === PropertiesOptions.PROPERTIES
              ? JSON.stringify(value)
              : value)
            : ''),
          type: fieldType
        };
      });
      return {
        refId: query.refId,
        fields: fields
      };
    }
    return {
      refId: query.refId,
      fields: []
    }
  }

  shouldRunQuery(query: ProductQuery): boolean {
    return true;
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    await this.get(this.baseUrl + '/v2/products?take=1');
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }

  async getFamilyNames(): Promise<void> {
    if (this.familyNamesCache.size > 0) {
      return;
    }

    const familyNames = await this.queryProductValues(PropertiesOptions.FAMILY)
      .catch(error => {
        this.error = parseErrorMessage(error)!;
      });

    familyNames?.forEach(familyName => this.familyNamesCache.set(familyName, familyName));
  }

  async metricFindQuery(query: ProductVariableQuery, options: LegacyMetricFindQueryOptions): Promise<MetricFindValue[]> {
    let metadata: ProductResponseProperties[];
    if (query.queryBy) {
      const filter = this.templateSrv.replace(query.queryBy, options.scopedVars)
      metadata = (await this.queryProducts(
        PropertiesOptions.PART_NUMBER,
        [Properties.partNumber, Properties.family],
        filter
      )).products;
    } else {
      metadata = (await this.queryProducts(
        PropertiesOptions.PART_NUMBER,
        [Properties.partNumber, Properties.family]
      )).products;
    }
    return metadata ? metadata.map(frame => ({ text: `${frame.partNumber}(${frame.family})`, value: frame.partNumber })) : [];
  }

  readonly productsComputedDataFields = new Map<string, ExpressionTransformFunction>(
    Object.values(ProductsQueryBuilderFieldNames).map(field => [
      field,
      field === ProductsQueryBuilderFieldNames.UPDATED_AT
        ? this.updatedAtQuery
        : this.multipleValuesQuery(field)
    ])
  );

  protected multipleValuesQuery(field: string): ExpressionTransformFunction {
    return (value: string, operation: string, _options?: any) => {
      const isMultiSelect = this.isMultiSelectValue(value);
      const valuesArray = this.getMultipleValuesArray(value);
      const logicalOperator = this.getLogicalOperator(operation);

      return isMultiSelect ? valuesArray
        .map(val => `${field} ${operation} "${val}"`)
        .join(` ${logicalOperator} `) : `${field} ${operation} "${value}"`;
    }
  }

  private getVariableOptions() {
    return this.templateSrv
      .getVariables()
      .map(variable => ({ label: '$' + variable.name, value: '$' + variable.name }));
  }

  private updatedAtQuery(value: string, operation: string): string {
    const formattedValue = value === '${__now:date}' ? new Date().toISOString() : value;
    return `${ProductsQueryBuilderFieldNames.UPDATED_AT} ${operation} "${formattedValue}"`;
  }

  private isMultiSelectValue(value: string): boolean {
    return value.startsWith('{') && value.endsWith('}');
  }

  private getMultipleValuesArray(value: string): string[] {
    return value.replace(/({|})/g, '').split(',');
  }

  private getLogicalOperator(operation: string): string {
    return operation === QueryBuilderOperations.EQUALS.name ? '||' : '&&';
  }

  private async loadWorkspaces(): Promise<void> {
    if (this.workspacesCache.size > 0) {
      return;
    }

    const workspaces = await this.getWorkspaces()
      .catch(error => {
        this.error = parseErrorMessage(error.message)!;
      });

    workspaces?.forEach(workspace => this.workspacesCache.set(workspace.id, workspace));

    this.workspacesLoaded();
  }

  private async getProductPartNumbers(): Promise<void> {
    if (this.partNumbersCache.size > 0) {
      return;
    }
    const partNumbers = await this.queryProductValues(PropertiesOptions.PART_NUMBER)
      .catch(error => {
        this.error = parseErrorMessage(error)!;
      });

    partNumbers?.forEach(partNumber => this.partNumbersCache.set(partNumber, partNumber));

    this.partNumberLoaded();
  }
}
