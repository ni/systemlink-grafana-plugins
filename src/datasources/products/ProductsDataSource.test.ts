import { BackendSrv } from '@grafana/runtime';
import { ProductsDataSource } from './ProductsDataSource';
import { ProductQuery, ProductVariableQuery, Properties, PropertiesOptions, QueryProductResponse } from './types';
import { Field, LegacyMetricFindQueryOptions } from '@grafana/data';
import { createFetchError, createFetchResponse, getQueryBuilder, requestMatching, setupDataSource } from 'test/fixtures';
import { MockProxy } from 'jest-mock-extended';
import { ProductsQueryBuilderFieldNames } from './constants/ProductsQueryBuilder.constants';
import { Workspace } from 'core/types';

const mockQueryProductResponse: QueryProductResponse = {
  products: [
    {
      id: '1',
      name: 'Product 1',
      partNumber: '123',
      family: 'Family 1',
      workspace: 'Workspace 1',
      updatedAt: '2021-08-01T00:00:00Z',
      properties: { prop1: 'value1' }
    }
  ],
  continuationToken: '',
  totalCount: 2
};

const mockVariableQueryProductResponse: QueryProductResponse = {
  products: [
    {
      id: '1',
      partNumber: '123',
      name: 'product 1',
    },
    {
      id: '2',
      partNumber: '456',
      name: 'product 2',
    }
  ],
  continuationToken: '',
  totalCount: 0
};

let datastore: ProductsDataSource, backendServer: MockProxy<BackendSrv>

beforeEach(() => {
  [datastore, backendServer] = setupDataSource(ProductsDataSource);

  backendServer.fetch
    .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
    .mockReturnValue(
      createFetchResponse<QueryProductResponse>(mockQueryProductResponse)
    );
});

describe('testDatasource', () => {
  test('returns success', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/products?take=1' }))
      .mockReturnValue(createFetchResponse(25));

    const result = await datastore.testDatasource();

    expect(result.status).toEqual('success');
  });

  test('bubbles up exception', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/products?take=1' }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.testDatasource())
      .rejects
      .toThrow('Request to url "/nitestmonitor/v2/products?take=1" failed with status code: 400. Error message: "Error"');
  });
});

describe('queryProducts', () => {
  test('returns data when there are valid queries', async () => {
    const response = await datastore.queryProducts();

    expect(response).toMatchSnapshot();
  });

  it('should throw error with status code when API returns error with status', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.queryProducts())
      .rejects
      .toThrow('The query failed due to the following error: (status 400) "Error".');
  });

  it('should throw error with unknown error when API returns error without status', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockImplementation(() => { throw new Error('Error'); });

    await expect(datastore.queryProducts())
      .rejects
      .toThrow('The query failed due to an unknown error.');
  });

  it('should publish alertError event when error occurs', async () => {
    const publishMock = jest.fn();
    (datastore as any).appEvents = { publish: publishMock };
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.queryProducts())
      .rejects
      .toThrow('The query failed due to the following error: (status 400) "Error".');

    expect(publishMock).toHaveBeenCalledWith({
      type: 'alert-error',
      payload: ['Error during product query', expect.stringContaining('The query failed due to the following error: (status 400) "Error".')],
    });
  });
});

describe('queryProductValues', () => {
  test('returns data when there are valid queries', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchResponse(['value1']));

    const response = await datastore.queryProductValues(ProductsQueryBuilderFieldNames.PART_NUMBER);

    expect(response).toMatchSnapshot();
  });

  test('raises an error when API fails', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchError(400));

    await expect(datastore.queryProductValues(ProductsQueryBuilderFieldNames.PART_NUMBER))
      .rejects
      .toThrow('Request to url "/nitestmonitor/v2/query-product-values" failed with status code: 400. Error message: "Error"');
  });
});

describe('getFamilyNames', () => {
  test('returns family names', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchResponse(['Family 1', 'Family 2']));

    await datastore.getFamilyNames();

    expect(datastore.familyNamesCache.get('Family 1')).toBe('Family 1');
    expect(datastore.familyNamesCache.get('Family 2')).toBe('Family 2');
  });

  test('should not query family values if cache exists', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchResponse(['value1']));
    datastore.familyNamesCache.set('family', 'value1');
    backendServer.fetch.mockClear();

    await datastore.getFamilyNames()

    expect(backendServer.fetch).not.toHaveBeenCalled();
  });

  it('should handle errors and set error and innerError fields', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchError(500));

    await datastore.getFamilyNames();

    expect(datastore.errorTitle).toBe('Warning during product value query');
    expect(datastore.errorDescription).toContain('Some values may not be available in the query builder lookups due to an unknown error.');
  });

  it('should handle errors and set innerError fields with error message detail', async () => {
    datastore.errorTitle = '';
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchError(500));

    await datastore.getFamilyNames();

    expect(datastore.errorTitle).toBe('Warning during product value query');
    expect(datastore.errorDescription).toContain('Some values may not be available in the query builder lookups due to the following error: \"Error\".');
  })

});

describe('query', () => {
  test('returns data when there are valid queries', async () => {
    const query = buildQuery(
      {
        refId: 'A',
        properties: [
          PropertiesOptions.PART_NUMBER,
          PropertiesOptions.FAMILY,
          PropertiesOptions.NAME,
          PropertiesOptions.WORKSPACE
        ] as Properties[],
        queryBy: `${ProductsQueryBuilderFieldNames.PART_NUMBER} = "123"`,
        orderBy: PropertiesOptions.ID,
        descending: false,
        recordCount: 1
      },
    );

    const response = await datastore.query(query);

    expect(response.data).toHaveLength(1);
    expect(response.data).toMatchSnapshot();
  });

  test('returns no data when Query Products returns no data', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockReturnValue(
        createFetchResponse(
          {
            products: [],
            continuationToken: null,
            totalCount: 0
          } as unknown as QueryProductResponse));

    const query = buildQuery();

    const response = await datastore.query(query);

    expect(response.data).toMatchSnapshot();
  });

  test('returns no data when Query Products returns error', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockReturnValue(createFetchError(400));

    const query = buildQuery(
      {
        refId: 'A',
        properties: [
          PropertiesOptions.PART_NUMBER,
          PropertiesOptions.FAMILY,
          PropertiesOptions.NAME,
          PropertiesOptions.WORKSPACE
        ] as Properties[], orderBy: undefined
      },
    );

    await expect(datastore.query(query))
      .rejects
      .toThrow('The query failed due to the following error: (status 400) \"Error\".');
  });

  it('should convert properties to Grafana fields', async () => {
    const query = buildQuery(
      {
        refId: 'A',
        properties: [
          PropertiesOptions.PART_NUMBER,
          PropertiesOptions.FAMILY,
          PropertiesOptions.NAME,
          PropertiesOptions.WORKSPACE,
          PropertiesOptions.UPDATEDAT,
          PropertiesOptions.PROPERTIES
        ] as Properties[], orderBy: undefined
      },
    );

    const response = await datastore.query(query);

    const fields = response.data[0].fields as Field[];
    expect(fields).toEqual([
      { name: 'partNumber', values: ['123'], type: 'string' },
      { name: 'family', values: ['Family 1'], type: 'string' },
      { name: 'name', values: ['Product 1'], type: 'string' },
      { name: 'workspace', values: ['Workspace 1'], type: 'string' },
      { name: 'updatedAt', values: ['2021-08-01T00:00:00Z'], type: 'time' },
      { name: 'properties', values: ['{"prop1":"value1"}'], type: 'string' },
    ]);
  });

  it('should convert workspaceIds to workspace names for workspace field', async () => {
    datastore.workspacesCache.set('Workspace 1', { id: 'Workspace 1', name: 'WorkspaceName'} as Workspace);

    const query = buildQuery(
      {
        refId: 'A',
        properties: [
          PropertiesOptions.WORKSPACE
        ] as Properties[], orderBy: undefined
      },
    );

    const response = await datastore.query(query);

    const fields = response.data[0].fields as Field[];
    expect(fields).toEqual([
      { name: 'workspace', values: ['WorkspaceName'], type: 'string' },
    ]);
  });

  test('should handle null and undefined properties', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
      .mockReturnValue(createFetchResponse({
        products: [
          {
            id: '1',
            name: 'Product 1',
            properties: null
          }
        ], continuationToken: null, totalCount: 0
      } as unknown as QueryProductResponse));

    const query = buildQuery(
      {
        refId: 'A',
        properties: [
          PropertiesOptions.PROPERTIES
        ] as Properties[], orderBy: undefined
      },
    );

    const response = await datastore.query(query);
    const fields = response.data[0].fields as Field[];
    expect(fields).toEqual([
      { name: 'properties', values: [''], type: 'string' },
    ]);
  });

  test('should not query product values if cache exists', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-product-values' }))
      .mockReturnValue(createFetchResponse(['value1']));
    datastore.partNumbersCache.set('partNumber', 'value1');
    backendServer.fetch.mockClear();

    await datastore.query(buildQuery())

    expect(backendServer.fetch).not.toHaveBeenCalled();
  });

  test('should not query workspace values if cache exists', async () => {
    backendServer.fetch
      .calledWith(requestMatching({ url: '/niauth/v1/user' }))
      .mockReturnValue(createFetchResponse(['workspace1']));
    datastore.workspacesCache.set('workspace', { id: 'workspace1', name: 'workspace1', default: false, enabled: true });
    backendServer.fetch.mockClear();

    await datastore.query(buildQuery())

    expect(backendServer.fetch).not.toHaveBeenCalled();
  });

  describe('Query builder queries', () => {
    test('should transform fields with single value', async () => {
      const query = buildQuery(
        {
          refId: 'A',
          properties: [
            PropertiesOptions.PART_NUMBER
          ] as Properties[],
          queryBy: `${PropertiesOptions.PART_NUMBER} = '123'`,
          descending: false
        },
      );

      await datastore.query(query);

      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: {
            descending: false,
            filter: "partNumber = '123'",
            orderBy: undefined,
            projection: ["partNumber"],
            returnCount: false, take: 1000
          }
        })
      );
    });

    test('should transform fields with multiple values', async () => {
      const query = buildQuery(
        {
          refId: 'A',
          properties: [
            PropertiesOptions.PART_NUMBER
          ] as Properties[],
          queryBy: `${ProductsQueryBuilderFieldNames.PART_NUMBER} = "{partNumber1,partNumber2}"`,
          descending: false
        },
      );

      await datastore.query(query);

      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: {
            descending: false,
            filter: "(PartNumber = \"partNumber1\" || PartNumber = \"partNumber2\")",
            orderBy: undefined,
            projection: ["partNumber"],
            returnCount: false, take: 1000
          }
        })
      );
    });

  });

  describe('metricFindQuery', () => {
    let options: LegacyMetricFindQueryOptions;
    beforeEach(() => {
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
        .mockReturnValue(
          createFetchResponse<QueryProductResponse>(mockVariableQueryProductResponse));
      options = {
        scopedVars: { partNumber: { value : '123' } }
      }
    });

    it('should return partNumber with family Name when queryBy is not provided', async () => {
      const query: ProductVariableQuery = {
        refId: '',
        queryBy: '',
      };

      const results = await datastore.metricFindQuery(query, options);

      expect(results).toMatchSnapshot();
      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: {
            descending: false,
            orderBy: "partNumber",
            projection: ["PART_NUMBER", "NAME"],
            returnCount: false,
          }
        })
      );
    });

    it('should return partNumber when queryBy is provided', async () => {
      const query: ProductVariableQuery = {
        refId: '',
        queryBy: 'partNumber = "123"',
      };

      const results = await datastore.metricFindQuery(query, options);

      expect(results).toMatchSnapshot();
      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: {
            descending: false,
            filter: "partNumber = \"123\"",
            orderBy: "partNumber",
            projection: ["PART_NUMBER", "NAME"],
            returnCount: false,
          }
        })
      );
    });

    it('should replace variables with values', async () => {
      const query: ProductVariableQuery = {
        refId: '',
        queryBy: 'partNumber = "$partNumber"',
      }

      const results = await datastore.metricFindQuery(query, options);

      expect(results).toMatchSnapshot();
      
    }); 

    it('should return empty array when queryBy is invalid', async () => {
      const query: ProductVariableQuery = {
        refId: '',
        queryBy: 'invalidQuery',
      };
      backendServer.fetch
        .calledWith(requestMatching({ url: '/nitestmonitor/v2/query-products' }))
        .mockReturnValue(
          createFetchResponse({
            products: [],
            continuationToken: null,
            totalCount: 0
          } as unknown as QueryProductResponse));
      options = {
        scopedVars: {}
      }

      const results = await datastore.metricFindQuery(query, options);

      expect(results).toEqual([]);
    })

    it('should handle multiple values in queryBy', async () => {
      const query: ProductVariableQuery = {
        refId: '',
        queryBy: `${ProductsQueryBuilderFieldNames.PART_NUMBER} = "{partNumber1,partNumber2}"`,
      };

      await datastore.metricFindQuery(query, options);

      expect(backendServer.fetch).toHaveBeenCalledWith(
        expect.objectContaining({
          data: {
            descending: false,
            filter:"(PartNumber = \"partNumber1\" || PartNumber = \"partNumber2\")",
            orderBy: "partNumber",
            projection: ["PART_NUMBER", "NAME"],
            returnCount: false,
          }
        })
      );
    });
  });
});

const buildQuery = getQueryBuilder<ProductQuery>()({
  refId: 'A',
  properties: [
    PropertiesOptions.PART_NUMBER,
    PropertiesOptions.FAMILY,
    PropertiesOptions.NAME,
    PropertiesOptions.WORKSPACE
  ] as Properties[],
  orderBy: undefined
});
