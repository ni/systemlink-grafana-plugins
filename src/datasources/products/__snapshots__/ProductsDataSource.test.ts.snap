// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query returns data when there are valid queries 1`] = `
[
  {
    "fields": [
      {
        "name": "partNumber",
        "type": "string",
        "values": [
          "123",
        ],
      },
      {
        "name": "family",
        "type": "string",
        "values": [
          "Family 1",
        ],
      },
      {
        "name": "name",
        "type": "string",
        "values": [
          "Product 1",
        ],
      },
      {
        "name": "workspace",
        "type": "string",
        "values": [
          "Workspace 1",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`query returns no data when Query Products returns no data 1`] = `[]`;

exports[`queryProductValues returns data when there are valid queries 1`] = `
[
  "value1",
]
`;

exports[`queryProducts returns data when there are valid queries 1`] = `
{
  "continuationToken": "",
  "products": [
    {
      "family": "Family 1",
      "id": "1",
      "name": "Product 1",
      "partNumber": "123",
      "properties": {
        "prop1": "value1",
      },
      "updatedAt": "2021-08-01T00:00:00Z",
      "workspace": "Workspace 1",
    },
  ],
  "totalCount": 2,
}
`;
